drop-transfer/src/ws/server/v4.rs

```diff
@@ -251,7 +331,33 @@ impl HandlerLoop<'_> {
     }
 
     async fn on_checksum(&mut self, report: v4::ReportChsum) {
-        if let Some(job) = self.jobs.get_mut(&report.file) {
+        let xfile = match self.xfer.files().get(&report.file) {
+            Some(file) => file,
+            None => return,
+        };
+
+        // Full checksum requsted at the begining of the transfer
+        if report.limit == xfile.size() {
+            self.checksums
+                .get(&report.file)
+                .expect("Missing file")
+                .or_set(report.checksum);
+
+            let storage = self.state.storage.clone();
+            let transfer_id = self.xfer.id();
+            let file_id = report.file.clone();
+            let logger = self.logger.clone();
+
+            tokio::spawn(async move {
+                if let Err(err) = storage
+                    .save_checksum(&transfer_id.to_string(), file_id.as_ref(), &report.checksum)
+                    .await
+                {
+                    error!(logger, "Failed to save checksum into DB: {err}");
+                }
+            });
+        // Requests made by the download task
+        } else if let Some(job) = self.jobs.get_mut(&report.file) {
```

Comment:
Yes, that is necessary if we want to preserve the protocol shape.
It should not be any race since we ask for the checksum in two places: 
* at the beginning we ask for a full checksum
* in case we're resuming the temp file, but in this case, we always ask for a partial checksum (since it is by definition not a full file)
so the `if report.limit == xfile.size()` should correctly differentiate this two cases